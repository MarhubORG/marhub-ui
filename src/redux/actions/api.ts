import {
  EXPORTING_IRAP_DATA,
  EXPORT_IRAP_DATA_FAILURE,
  EXPORT_IRAP_DATA_SUCCESS,
  EXPORTING_BOTPRESS_DATA,
  EXPORT_BOTPRESS_DATA_FAILURE,
  EXPORT_BOTPRESS_DATA_SUCCESS,
  LOGOUT,
} from '../constants/actionTypes';

export interface ExportingIrapDataPayload {
  startDate: string;
  endDate: string;
  selectedTemplate: string;
  emailText: string;
  irapUuidSearchText: string;
}

export interface ExportingIrapDataAction {
  type: typeof EXPORTING_IRAP_DATA;
  payload: ExportingIrapDataPayload;
}

export function exportingIrapData(
  data: ExportingIrapDataPayload
): ExportingIrapDataAction {
  return { type: EXPORTING_IRAP_DATA, payload: data };
}

export interface ExportIrapDataSuccessAction {
  type: typeof EXPORT_IRAP_DATA_SUCCESS;
  payload: object[];
}

export function exportingIrapDataSuccess(
  data: object[]
): ExportIrapDataSuccessAction {
  return {
    type: EXPORT_IRAP_DATA_SUCCESS,
    payload: data,
  };
}

export interface ExportIrapDataFailureAction {
  type: typeof EXPORT_IRAP_DATA_FAILURE;
  payload: string;
}

export function exportingIrapDataFailure(
  message: string
): ExportIrapDataFailureAction {
  return {
    type: EXPORT_IRAP_DATA_FAILURE,
    payload: message,
  };
}

export interface ExportingBotpressDataPayload {
  startDate: string;
  endDate: string;
  selectedTemplate: string;
  emailText: string;
  irapUuidSearchText: string;
}

export interface ExportingBotpressDataAction {
  type: typeof EXPORTING_BOTPRESS_DATA;
  payload: ExportingBotpressDataPayload;
}

export function exportingBotpressData(
  data: ExportingBotpressDataPayload
): ExportingBotpressDataAction {
  return { type: EXPORTING_BOTPRESS_DATA, payload: data };
}

export interface ExportBotpressDataSuccessAction {
  type: typeof EXPORT_BOTPRESS_DATA_SUCCESS;
  payload: object[];
}

export function exportingBotpressDataSuccess(
  data: object[]
): ExportBotpressDataSuccessAction {
  return {
    type: EXPORT_BOTPRESS_DATA_SUCCESS,
    payload: data,
  };
}

export interface ExportBotpressDataFailureAction {
  type: typeof EXPORT_BOTPRESS_DATA_FAILURE;
  payload: string;
}

export function exportingBotpressDataFailure(
  message: string
): ExportBotpressDataFailureAction {
  return {
    type: EXPORT_BOTPRESS_DATA_FAILURE,
    payload: message,
  };
}

export interface LogoutAction {
  type: typeof LOGOUT;
}

export function logout(): LogoutAction {
  return { type: LOGOUT };
}

export type ApiActionTypes =
  | ExportingIrapDataAction
  | ExportIrapDataSuccessAction
  | ExportIrapDataFailureAction
  | ExportingBotpressDataAction
  | ExportBotpressDataSuccessAction
  | ExportBotpressDataFailureAction
  | LogoutAction;
