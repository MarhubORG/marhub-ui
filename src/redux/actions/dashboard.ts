import {
  FETCH_ORGANIZATIONS,
  FETCH_ORGANIZATIONS_FAILURE,
  FETCH_ORGANIZATIONS_SUCCESS,
  LOGOUT,
  UPDATE_ORGANIZATION,
  UPDATE_ORGANIZATION_FAILURE,
  UPDATE_ORGANIZATION_SUCCESS,
} from '../constants/actionTypes';

export interface FetchOrganizationsAction {
  type: typeof FETCH_ORGANIZATIONS;
}

export function fetchOrganizations(): FetchOrganizationsAction {
  return { type: FETCH_ORGANIZATIONS };
}

export interface OrganizationFields {
  name: string;
  visible_fields: string[];
}

export interface Organization {
  organisation: OrganizationFields;
  id: number;
}

export interface FetchOrganizationsSuccessAction {
  type: typeof FETCH_ORGANIZATIONS_SUCCESS;
  payload: Organization[];
}

export function fetchOrganizationsSuccess(
  organizations: Organization[]
): FetchOrganizationsSuccessAction {
  return { type: FETCH_ORGANIZATIONS_SUCCESS, payload: organizations };
}

export interface FetchOrganizationsFailureAction {
  type: typeof FETCH_ORGANIZATIONS_FAILURE;
}

export function fetchOrganizationsFailure(): FetchOrganizationsFailureAction {
  return { type: FETCH_ORGANIZATIONS_FAILURE };
}

export interface LogoutAction {
  type: typeof LOGOUT;
}

export function logout(): LogoutAction {
  return { type: LOGOUT };
}

export interface UpdateOrganizationPayload {
  id: number;
  organization: Organization;
}

export interface UpdateOrganizationAction {
  type: typeof UPDATE_ORGANIZATION;
  payload: UpdateOrganizationPayload;
}

export function updateOrganization(
  organization: UpdateOrganizationPayload
): UpdateOrganizationAction {
  return {
    type: UPDATE_ORGANIZATION,
    payload: organization,
  };
}

export interface UpdateOrganizationSuccessPayload {
  id: number;
  organization: Organization;
}

export interface UpdateOrganizationSuccessAction {
  type: typeof UPDATE_ORGANIZATION_SUCCESS;
  payload: UpdateOrganizationSuccessPayload;
}

export function updateOrganizationSuccess(
  organization: UpdateOrganizationSuccessPayload
): UpdateOrganizationSuccessAction {
  return {
    type: UPDATE_ORGANIZATION_SUCCESS,
    payload: organization,
  };
}

export interface UpdateOrganizationFailureAction {
  type: typeof UPDATE_ORGANIZATION_FAILURE;
  payload: string;
}

export function updateOrganizationFailure(
  message: string
): UpdateOrganizationFailureAction {
  return {
    type: UPDATE_ORGANIZATION_FAILURE,
    payload: message,
  };
}

export type DashboardActionTypes =
  | LogoutAction
  | FetchOrganizationsAction
  | FetchOrganizationsSuccessAction
  | FetchOrganizationsFailureAction
  | UpdateOrganizationAction
  | UpdateOrganizationSuccessAction
  | UpdateOrganizationFailureAction;
