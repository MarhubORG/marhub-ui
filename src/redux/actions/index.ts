import { SIGNUP, SIGNUP_REDIRECTING } from '../constants/actionTypes';
import { Action } from '../../types/interfaces';

export function signup(
  name: string,
  organization: string,
  email: string,
  password: string
): Action {
  return { type: SIGNUP, payload: { name, organization, email, password } };
}

export function signupRedirecting(): Action {
  return { type: SIGNUP_REDIRECTING };
}
