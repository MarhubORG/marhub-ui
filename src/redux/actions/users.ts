import {
  FETCH_USERS,
  FETCH_USERS_FAILURE,
  FETCH_USERS_SUCCESS,
  CREATE_USER,
  CREATE_USER_FAILURE,
  CREATE_USER_SUCCESS,
  EDIT_USER,
  EDIT_USER_FAILURE,
  EDIT_USER_SUCCESS,
  DELETE_USER,
  DELETE_USER_FAILURE,
  DELETE_USER_SUCCESS,
  LOGOUT,
} from '../constants/actionTypes';

export interface FetchUsersAction {
  type: typeof FETCH_USERS;
}

export interface UserProfileFields {
  name: string;
  organisation: string;
  role?: string;
}

export interface User {
  id: number;
  email: string;
  isDisabled: boolean;
  profile: UserProfileFields;
}

export function fetchUsers(): FetchUsersAction {
  return {
    type: FETCH_USERS,
  };
}

export interface FetchUsersFailureAction {
  type: typeof FETCH_USERS_FAILURE;
}

export function fetchUsersFailure(): FetchUsersFailureAction {
  return {
    type: FETCH_USERS_FAILURE,
  };
}

export interface FetchUsersSuccessAction {
  type: typeof FETCH_USERS_SUCCESS;
  payload: User[];
}

export function fetchUsersSuccess(payload: User[]): FetchUsersSuccessAction {
  return {
    type: FETCH_USERS_SUCCESS,
    payload,
  };
}

export interface CreateUserPayload {
  email: string;
  name: string;
  selectedOrganization: string;
  role: string;
  password: string;
  isDisabled: boolean;
}

export interface EditUserPayload {
  email: string;
  name: string;
  selectedOrganization: string;
  role: string;
  password: string;
  isDisabled: boolean;
  id: number;
}

export interface CreateUserAction {
  type: typeof CREATE_USER;
  payload: CreateUserPayload;
}

export function createUser(data: CreateUserPayload): CreateUserAction {
  return {
    type: CREATE_USER,
    payload: {
      email: data.email,
      name: data.name,
      selectedOrganization: data.selectedOrganization,
      role: data.role,
      password: data.password,
      isDisabled: data.isDisabled,
    },
  };
}

export interface CreateUserFailureAction {
  type: typeof CREATE_USER_FAILURE;
}

export function createUserFailure(): CreateUserFailureAction {
  return {
    type: CREATE_USER_FAILURE,
  };
}

export interface CreateUserSuccessAction {
  type: typeof CREATE_USER_SUCCESS;
  payload: User;
}

export function createUserSuccess(payload: User): CreateUserSuccessAction {
  return {
    type: CREATE_USER_SUCCESS,
    payload,
  };
}

export interface EditUserAction {
  type: typeof EDIT_USER;
  payload: EditUserPayload;
}

export function editUser(payload: EditUserPayload): EditUserAction {
  return {
    type: EDIT_USER,
    payload,
  };
}

export interface EditUserSuccessAction {
  type: typeof EDIT_USER_SUCCESS;
  payload: User;
}

export function editUserSuccess(payload: User): EditUserSuccessAction {
  return {
    type: EDIT_USER_SUCCESS,
    payload,
  };
}

export interface EditUserFailureAction {
  type: typeof EDIT_USER_FAILURE;
}

export function editUserFailure(): EditUserFailureAction {
  return {
    type: EDIT_USER_FAILURE,
  };
}

export interface LogoutAction {
  type: typeof LOGOUT;
}

export function logout(): LogoutAction {
  return {
    type: LOGOUT,
  };
}

export interface DeleteUserAction {
  type: typeof DELETE_USER;
  payload: number;
}

export function deleteUser(payload: number): DeleteUserAction {
  return {
    type: DELETE_USER,
    payload,
  };
}

export interface DeleteUserFailureAction {
  type: typeof DELETE_USER_FAILURE;
}

export function deleteUserFailure(): DeleteUserFailureAction {
  return {
    type: DELETE_USER_FAILURE,
  };
}

export interface DeleteUserSuccessAction {
  type: typeof DELETE_USER_SUCCESS;
}

export function deleteUserSuccess(): DeleteUserSuccessAction {
  return {
    type: DELETE_USER_SUCCESS,
  };
}

export type UserActionTypes =
  | FetchUsersAction
  | FetchUsersFailureAction
  | FetchUsersSuccessAction
  | CreateUserAction
  | CreateUserFailureAction
  | CreateUserSuccessAction
  | EditUserAction
  | EditUserSuccessAction
  | EditUserFailureAction
  | DeleteUserAction
  | DeleteUserFailureAction
  | DeleteUserSuccessAction
  | LogoutAction;
