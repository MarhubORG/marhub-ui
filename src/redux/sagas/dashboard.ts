import axios from 'axios';
import { put, takeLatest } from 'redux-saga/effects';
import cookie from 'react-cookies';

import {
  FETCH_ORGANIZATIONS,
  UPDATE_ORGANIZATION,
  CREATE_ORGANIZATION,
  CREATE_TEMPLATE,
  UPDATE_TEMPLATE,
  DELETE_TEMPLATE,
} from '../constants/actionTypes';
import {
  fetchOrganizationsSuccess,
  fetchOrganizationsFailure,
  FetchOrganizationsAction,
  UpdateOrganizationAction,
  updateOrganizationSuccess,
  updateOrganizationFailure,
  CreateOrganizationAction,
  createOrganizationFailure,
  createOrganizationSuccess,
  CreateTemplateAction,
  createTemplateFailure,
  createTemplateSuccess,
  UpdateTemplateAction,
  updateTemplateSuccess,
  updateTemplateFailure,
  DeleteTemplateAction,
  deleteTemplateSuccess,
  deleteTemplateFailure,
} from '../actions/dashboard';

function* fetchOrgs(action: FetchOrganizationsAction): object {
  try {
    const token = cookie.load('token');
    const axiosConfig = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const url = 'http://localhost:8080/api/v1/organisations';

    const json = yield axios.get(url, axiosConfig);
    yield put(fetchOrganizationsSuccess(json.data[0]));
  } catch (error) {
    // const errorMessage = `${error.request.status} Error: ${error.response.data.error}`;
    yield put(fetchOrganizationsFailure());
  }
}

export function* fetchOrganizationsWatcher(): object {
  yield takeLatest(FETCH_ORGANIZATIONS, fetchOrgs);
}

function* updateOrganization(action: UpdateOrganizationAction): object {
  try {
    const token = cookie.load('token');
    const url = 'http://localhost:8080/api/v1/exportation_fields';
    axios.defaults.headers.common.Authorization = `Bearer ${token}`;
    const json = yield axios.post(url, {
      id: action.payload.id,
      organization: action.payload.organization,
    });
    yield put(updateOrganizationSuccess(json.data));
  } catch (error) {
    yield put(
      updateOrganizationFailure(
        'Update organization failed. Please try again or contact your administrator.'
      )
    );
  }
}

export function* updateOrganizationWatcher(): object {
  yield takeLatest(UPDATE_ORGANIZATION, updateOrganization);
}

function* createOrganization(action: CreateOrganizationAction): object {
  try {
    const token = cookie.load('token');
    const url = 'http://localhost:8080/api/v1/organisations';
    axios.defaults.headers.common.Authorization = `Bearer ${token}`;
    const json = yield axios.post(url, {
      name: action.payload,
    });
    yield put(createOrganizationSuccess(json.data));
  } catch (error) {
    yield put(createOrganizationFailure());
  }
}

export function* createOrganizationWatcher(): object {
  yield takeLatest(CREATE_ORGANIZATION, createOrganization);
}

function* createTemplate(action: CreateTemplateAction): object {
  try {
    const token = cookie.load('token');
    const url = 'http://localhost:8080/api/v1/templates';
    axios.defaults.headers.common.Authorization = `Bearer ${token}`;
    const json = yield axios.post(url, {
      name: action.payload.name,
      fields: action.payload.fields,
    });
    yield put(createTemplateSuccess(json.data));
  } catch (error) {
    yield put(createTemplateFailure());
  }
}

export function* createTemplateWatcher(): object {
  yield takeLatest(CREATE_TEMPLATE, createTemplate);
}

function* updateTemplate(action: UpdateTemplateAction): object {
  try {
    const token = cookie.load('token');
    const url = 'http://localhost:8080/api/v1/templates';
    axios.defaults.headers.common.Authorization = `Bearer ${token}`;
    const json = yield axios.post(url, {
      name: action.payload.name,
      fields: action.payload.fields,
    });
    yield put(updateTemplateSuccess(json.data));
  } catch (error) {
    const errorMessage =
      'Update template failed. Please contact the administrator.';
    yield put(updateTemplateFailure(errorMessage));
  }
}

export function* updateTemplateWatcher(): object {
  yield takeLatest(UPDATE_TEMPLATE, updateTemplate);
}

function* deleteTemplate(action: DeleteTemplateAction): object {
  try {
    const token = cookie.load('token');
    const url = 'http://localhost:8080/api/v1/templates';
    axios.defaults.headers.common.Authorization = `Bearer ${token}`;
    const json = yield axios.delete(url, {
      data: {
        name: action.payload,
      },
    });
    yield put(deleteTemplateSuccess(json.data));
  } catch (error) {
    yield put(
      deleteTemplateFailure(
        'Delete template failed. Please contact the administrator.'
      )
    );
  }
}

export function* deleteTemplateWatcher(): object {
  yield takeLatest(DELETE_TEMPLATE, deleteTemplate);
}
