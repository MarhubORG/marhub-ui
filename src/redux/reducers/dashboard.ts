import { DashboardState } from '../../types/interfaces';
import {
  FETCH_ORGANIZATIONS,
  FETCH_ORGANIZATIONS_FAILURE,
  FETCH_ORGANIZATIONS_SUCCESS,
  UPDATE_ORGANIZATION,
  UPDATE_ORGANIZATION_FAILURE,
  UPDATE_ORGANIZATION_SUCCESS,
  LOGOUT,
  CREATE_ORGANIZATION,
  CREATE_ORGANIZATION_FAILURE,
  CREATE_ORGANIZATION_SUCCESS,
  CREATE_ORGANIZATION_REDIRECT,
  CREATE_TEMPLATE,
  CREATE_TEMPLATE_FAILURE,
  CREATE_TEMPLATE_SUCCESS,
  UPDATE_TEMPLATE,
  UPDATE_TEMPLATE_FAILURE,
  UPDATE_TEMPLATE_SUCCESS,
  DELETE_TEMPLATE,
  DELETE_TEMPLATE_FAILURE,
  DELETE_TEMPLATE_SUCCESS,
  DELETE_TEMPLATE_REDIRECT,
} from '../constants/actionTypes';
import { DashboardActionTypes, Organization } from '../actions/dashboard';

export const standardFetchOrganizationErrorMessage =
  'Could not fetch organizations. Please try again later or contact the administrator.';
export const templateFailureMessage =
  'Creating template failed. Please contact your administrator.';
export const templateSuccessMessage = 'Successfully created new template';

export const initialState = {
  loading: false,
  organizations: [],
  errorMessage: '',
  redirectToVisibleFields: '',
  templateMessage: '',
};
export const successMessage = 'Successfully updated organization permissions.';
export const createOrgFailureMessage =
  'Creating a new organization failed. Please contact your administrator.';
export const createOrgSuccessMessage =
  'Successfully updated organization permissions.';
export default function dashboardReducer(
  state = initialState,
  action: DashboardActionTypes
): DashboardState {
  switch (action.type) {
    case FETCH_ORGANIZATIONS:
      return { ...state, loading: true, errorMessage: '' };
    case FETCH_ORGANIZATIONS_SUCCESS:
      return {
        ...state,
        loading: false,
        organizations: [...action.payload],
        errorMessage: '',
      };
    case FETCH_ORGANIZATIONS_FAILURE:
      return {
        ...state,
        loading: false,
        errorMessage: standardFetchOrganizationErrorMessage,
      };
    case UPDATE_ORGANIZATION:
      return { ...state, loading: true, errorMessage: '' };
    case UPDATE_ORGANIZATION_FAILURE:
      return { ...state, loading: false, errorMessage: action.payload };
    case UPDATE_ORGANIZATION_SUCCESS:
      return {
        ...state,
        loading: false,
        errorMessage: successMessage,
        organizations: replaceWithUpdatedOrg(action.payload, state),
      };
    case LOGOUT:
      return { ...initialState };
    case CREATE_ORGANIZATION:
      return {
        ...state,
        loading: true,
        errorMessage: '',
      };
    case CREATE_ORGANIZATION_FAILURE:
      return {
        ...state,
        loading: false,
        errorMessage: createOrgFailureMessage,
      };
    case CREATE_ORGANIZATION_SUCCESS:
      return {
        ...state,
        organizations: [...state.organizations, action.payload],
        loading: false,
        errorMessage: createOrgSuccessMessage,
        redirectToVisibleFields: action.payload.organisation.name,
      };
    case CREATE_ORGANIZATION_REDIRECT:
      return {
        ...state,
        redirectToVisibleFields: '',
      };
    case CREATE_TEMPLATE_FAILURE:
      return {
        ...state,
        templateMessage: templateFailureMessage,
        loading: false,
      };
    case CREATE_TEMPLATE:
      return {
        ...state,
        loading: true,
        templateMessage: '',
      };
    case CREATE_TEMPLATE_SUCCESS:
      return {
        ...state,
        loading: false,
        templateMessage: templateSuccessMessage,
        organizations: replaceWithUpdatedOrg(action.payload, state),
      };
    case UPDATE_TEMPLATE:
      return {
        ...state,
        loading: true,
        templateMessage: '',
      };
    case UPDATE_TEMPLATE_FAILURE:
      return {
        ...state,
        loading: false,
        templateMessage: action.payload,
      };
    case UPDATE_TEMPLATE_SUCCESS:
      return {
        ...state,
        loading: false,
        templateMessage: 'Updated template successfully',
        organizations: replaceWithUpdatedOrg(action.payload, state),
      };
    case DELETE_TEMPLATE:
      return {
        ...state,
        loading: true,
        templateMessage: '',
      };
    case DELETE_TEMPLATE_FAILURE:
      return {
        ...state,
        loading: false,
        templateMessage: action.payload,
      };
    case DELETE_TEMPLATE_SUCCESS:
      return {
        ...state,
        loading: false,
        templateMessage: 'Template Deleted.',
        organizations: replaceWithUpdatedOrg(action.payload, state),
      };
    case DELETE_TEMPLATE_REDIRECT:
      return {
        ...state,
        loading: false,
        templateMessage: '',
      };
    default:
      return state;
  }
}

export function replaceWithUpdatedOrg(
  organization: Organization,
  state: DashboardState
): Organization[] {
  return state.organizations.map(el => {
    if (el.id === organization.id) return organization;
    return el;
  });
}
