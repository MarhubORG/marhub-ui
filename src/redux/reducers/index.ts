import { combineReducers } from 'redux';
import registration from './registration';
import apiReducer from './api';
import dashboardReducer from './dashboard';

export default combineReducers({
  registration,
  apiReducer,
  dashboardReducer,
});
