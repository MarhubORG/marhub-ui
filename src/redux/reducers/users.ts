import { UserState } from '../../types/interfaces';
import { UserActionTypes, User } from '../actions/users';
import {
  FETCH_USERS,
  FETCH_USERS_FAILURE,
  FETCH_USERS_SUCCESS,
  CREATE_USER,
  CREATE_USER_FAILURE,
  CREATE_USER_SUCCESS,
  EDIT_USER,
  EDIT_USER_FAILURE,
  EDIT_USER_SUCCESS,
  DELETE_USER,
  DELETE_USER_FAILURE,
  DELETE_USER_SUCCESS,
  LOGOUT,
} from '../constants/actionTypes';

export const failureMessage =
  'Loading Users failed. Please contact your administrator.';
export const createUserFailureMessage =
  'Creating user failed. Please contact your administrator.';
export const editUserFailureMessage =
  'Editing user failed. Please contact your administrator.';
export const editUserSuccessMessage =
  'Edit user success! All changes have been saved to the database.';

export const initialState = {
  users: [],
  message: '',
  loading: false,
};
export default function userReducer(
  state = initialState,
  action: UserActionTypes
): UserState {
  switch (action.type) {
    case FETCH_USERS:
      return { ...state, loading: true, message: '' };
    case FETCH_USERS_FAILURE:
      return { ...state, loading: false, message: failureMessage };
    case FETCH_USERS_SUCCESS:
      return { ...state, loading: false, message: '', users: action.payload };
    case CREATE_USER:
      return { ...state, loading: true, message: '' };
    case CREATE_USER_FAILURE:
      return { ...state, loading: false, message: createUserFailureMessage };
    case CREATE_USER_SUCCESS:
      return {
        ...state,
        loading: false,
        message: 'CREATE_USER_SUCCESS',
        users: [...state.users, action.payload],
      };
    case EDIT_USER:
      return {
        ...state,
        loading: true,
        message: '',
      };
    case EDIT_USER_FAILURE:
      return {
        ...state,
        loading: false,
        message: editUserFailureMessage,
      };
    case EDIT_USER_SUCCESS:
      return {
        ...state,
        loading: false,
        message: editUserSuccessMessage,
        users: replaceWithUpdatedUser(action.payload, state),
      };
    case DELETE_USER:
      return {
        ...state,
        loading: true,
        message: '',
      };
    case DELETE_USER_SUCCESS:
      return {
        ...state,
        loading: false,
        message: 'DELETE_USER_SUCCESS',
      };
    case DELETE_USER_FAILURE:
      return {
        ...state,
        loading: false,
        message: 'Delete user failed. Please contact the administrator.',
      };
    case LOGOUT:
      return initialState;
    default:
      return state;
  }
}

export function replaceWithUpdatedUser(user: User, state: UserState): User[] {
  return state.users.map(el => {
    if (el.id === user.id) return user;
    return el;
  });
}
