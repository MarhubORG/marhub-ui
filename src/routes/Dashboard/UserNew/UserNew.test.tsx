import React from 'react';
import renderer from 'react-test-renderer';
import {
  UnconnectedUserNew,
  MapDispatchtoProps,
  mapDispatchToProps,
  mapStateToProps,
} from './UserNew';
import { fetchOrganizations } from '../../../redux/actions/dashboard';
import { createUser } from '../../../redux/actions/users';
import { initialState } from '../../../redux/reducers/dashboard';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <UnconnectedUserNew
        fetchOrganizations={fetchOrganizations}
        organizations={[]}
        createUser={createUser}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('mapDispatchToProps', () => {
  it('contains functions', () => {
    const dispatch = jest.fn();
    const dispatches: MapDispatchtoProps = mapDispatchToProps(dispatch);
    expect(Object.keys(dispatches).includes('fetchOrganizations')).toBeTruthy();
    expect(Object.keys(dispatches).includes('createUser')).toBeTruthy();
    expect(
      dispatches.createUser({
        email: 'email',
        name: 'name',
        selectedOrganization: '1',
        role: 'role',
        password: 'password',
        isDisabled: false,
      })
    ).toBeUndefined();
    expect(dispatches.fetchOrganizations()).toBeUndefined();
  });
});
