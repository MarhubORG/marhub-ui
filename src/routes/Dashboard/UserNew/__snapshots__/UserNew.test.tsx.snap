// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c3 {
  height: 2rem;
  min-width: 18rem;
  max-width: 18rem;
  border-radius: 0.2rem;
  font-size: 0.9rem;
  padding-left: 1rem;
  padding-right: 1rem;
  border: 1px solid #edf1f7;
}

.c2 {
  display: block;
  font-family: Open Sans,sans-serif;
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  margin: 0.5rem 0rem;
}

.c4 {
  display: block;
  font-family: Open Sans,sans-serif;
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  margin: 0.5rem 0rem;
}

.c5 {
  height: 2rem;
  min-width: 20.2rem;
  max-width: 20.2rem;
  border-radius: 0.2rem;
  font-size: 0.85rem;
  padding-left: 1rem;
  padding-right: 1rem;
  border: 1px solid #edf1f7;
}

.c6 {
  margin-left: 2rem;
}

.c0 {
  margin: 1rem;
}

.c7 {
  display: block;
  font-family: Open Sans,sans-serif;
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  margin: 0.5rem 0rem;
}

.c8 {
  height: 2rem;
  width: 8rem;
  border-radius: 5px;
  font-size: 0.9rem;
}

.c1 {
  margin-left: -1rem;
}

<div
  className="c0"
>
  <h1>
    New User
  </h1>
  <div
    className="c1"
  />
  <div>
    <label
      className="c2"
      htmlFor="name"
    >
      Name: *
    </label>
    <input
      className="c3"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      className="c2"
      htmlFor="email"
    >
      Email: *
    </label>
    <input
      className="c3"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      className="c2"
      htmlFor="password"
    >
      Password: *
    </label>
    <input
      className="c3"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      className="c4"
      htmlFor="Select Organization: *"
    >
      Select Organization: *
    </label>
    <select
      className="c5"
      id="Select Organization: *"
      onChange={[Function]}
    >
      <option
        className="c6"
      >
        Please Select:
      </option>
    </select>
  </div>
  <div>
    <label
      className="c4"
      htmlFor="Select Role: *"
    >
      Select Role: *
    </label>
    <select
      className="c5"
      id="Select Role: *"
      onChange={[Function]}
    >
      <option
        className="c6"
      >
        Please Select:
      </option>
      <option
        className="c6"
        selected={false}
        value="user"
      >
        user
      </option>
      <option
        className="c6"
        selected={false}
        value="admin"
      >
        admin
      </option>
      <option
        className="c6"
        selected={false}
        value="marhub_user"
      >
        marhub_user
      </option>
      <option
        className="c6"
        selected={false}
        value="marhub_admin"
      >
        marhub_admin
      </option>
    </select>
  </div>
  <label
    className="c7"
    htmlFor="isDisabled"
  >
    Account is Disabled:
    <input
      checked={false}
      id="isDisabled"
      name="isDisabled"
      onChange={[Function]}
      type="checkbox"
    />
  </label>
  <button
    className="c8"
    onClick={[Function]}
    type="button"
  >
    Submit
  </button>
</div>
`;
